lib/ansible/modules/network/frr/frr_facts.py

```diff
@@ -0,0 +1,470 @@
+        self.module = module
+        self.facts = dict()
+        self.responses = None
+        self._capabilities = get_capabilities(self.module)
+
+    def populate(self):
+        self.responses = run_commands(self.module, commands=self.COMMANDS, check_rc=False)
+
+    def run(self, cmd):
+        return run_commands(commands=cmd, check_rc=False)
+
+
+class Default(FactsBase):
+
+    COMMANDS = ['show version']
+
+    def populate(self):
+        super(Default, self).populate()
+        self.facts.update(self.platform_facts())
+
+    def platform_facts(self):
+        platform_facts = {}
+
+        resp = self._capabilities
+        device_info = resp['device_info']
+
+        platform_facts['system'] = device_info['network_os']
+
+        for item in ('version', 'hostname'):
+            val = device_info.get('network_os_%s' % item)
+            if val:
+                platform_facts[item] = val
+
+        return platform_facts
+
+
+class Hardware(FactsBase):
+
+    COMMANDS = ['show memory']
+
+    def _parse_total_heap_allocated(self, data):
+        match = re.search(r'Total heap allocated:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_holding_block_headers(self, data):
+        match = re.search(r'Holding block headers:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_used_small_blocks(self, data):
+        match = re.search(r'Used small blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_used_ordinary_blocks(self, data):
+        match = re.search(r'Used ordinary blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_free_small_blocks(self, data):
+        match = re.search(r'Free small blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_free_ordinary_blocks(self, data):
+        match = re.search(r'Free ordinary blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_ordinary_blocks(self, data):
+        match = re.search(r'Ordinary blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_small_blocks(self, data):
+        match = re.search(r'Small blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_holding_blocks(self, data):
+        match = re.search(r'Holding blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_daemons(self, data):
+        match = re.search(r'Memory statistics for (\w+)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def gather_memory_facts(self, data):
+        mem_details = data.split('\n\n')
+        mem_stats = {}
+        mem_counters = ['total_heap_allocated', 'holding_block_headers', 'used_small_blocks',
```

Comment:
This can be a key-value map with key as the fact name and value as the pattern to be searched for